Make progress bar available with n_jobs!=1

masap opened the pull request:
## Motivation
Make progress bar available with ```n_jobs!=1```.

## Description of the changes
This PR fixes issue #2957.

Signed-off-by: Masashi Honma <masashi.honma@gmail.com>

HideakiImamura commented:
@toshihikoyanase @himkt Could you review this PR?

@xadrianzetx Please feel free to review this PR if you have time as an author of #3115.

HideakiImamura commented:
@masap Let me share the useful information from @xadrianzetx. You can check https://github.com/optuna/optuna/pull/3115#discussion_r760397095 to try the progress bar with `n_jobs!=1`.

masap commented:
@HideakiImamura Thanks! I will try it tomorrow.

masap commented:
> Let me share the useful information from @xadrianzetx. You can check #3115 (comment) to try the progress bar with n_jobs!=1.

Great ! It works with my patch also.

masap commented:
I fixed the PR. The previous patch made all processes outputting a progrress bar to a single ```stderr```. Since each process may show the progress ```100%``` twice, counting ```100%``` in single ```stderr``` could not find out a process which did now show ```100%```. The new patch will check if ```100%``` is displayed for each process.

masap commented:
I am facing a RDB stroage failure. But I can not reproduce it in my local environment even if I use same python version... So I am testing on this CI. Sorry for frequent resubmit.

masap commented:
These tests are running more than 2 hours unexpectedly. It runs about 3minutes in ordinary case. Could anybody stop the jobs ?

```
Tests (RDB storage) / tests-rdbstorage (3.6) (pull_request)
Tests (RDB storage) / tests-rdbstorage (3.7) (pull_request)
```

himkt commented:
No worries @masap and I force stopped jobs.
As a side note, you can run CIs with your forked repo (open pull request from the forked repo to the forked repo).
Of course, there is no problem to run CIs on this PR but you can cancel jobs with your permission if jobs run on the forked repo.

masap commented:
@himkt Thank you! I will try it on my repo.

HideakiImamura commented:
@nzw0301 Could you review this PR instead of @himkt? I discussed it with him privately and came to the conclusion that it would be better to change reviewers.

nzw0301 commented:
sure.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -26,13 +29,33 @@ def objective(trial: optuna.Trial) -> float:
 
 
-def run_optimize(args: Tuple[str, str]) -> None:
+def run_optimize(args: Tuple[str, str, int, bool]) -> str:
```
I prefer defining another function for the progress bar to extend this `run_optimize`. What do you think?

masap replied:
Ok, I defined another function.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -107,11 +130,57 @@ def test_multiprocess(storage_url: str) -> None:
     _check_trials(trials)
+
+
+def test_multiprocess_with_progbar(
```
```suggestion
@pytest.mark.parametrize("n_jobs", [1, 4])
def test_multiprocess_with_progbar(
```

masap replied:
Ok, fixed.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -107,11 +130,57 @@ def test_multiprocess(storage_url: str) -> None:
+
+
+def test_multiprocess_with_progbar(
+    capsys: _pytest.capture.CaptureFixture, storage_url: str
```
```suggestion
    capsys: _pytest.capture.CaptureFixture, storage_url: str, n_jobs: int
```

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -107,11 +130,57 @@ def test_multiprocess(storage_url: str) -> None:
+        study_name = _STUDY_NAME
+        optuna.create_study(storage=storage_url, study_name=study_name)
+        with Pool(n_workers) as pool:
+            ret = pool.map(run_optimize, [(study_name, storage_url, 1, True)] * n_workers)
```
```suggestion
            ret = pool.map(run_optimize, [(study_name, storage_url, n_jobs, True)] * n_workers)
```

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -107,11 +130,57 @@ def test_multiprocess(storage_url: str) -> None:
+        _check_trials(trials)
+
+
+def test_multiprocess_with_progbar_parallel(
```
We can merge this function to `test_multiprocess_with_progbar` with `parametrize` decorator.

masap commented:
In addition, I fixed a warning from black to avoid CI failure.

masap commented:
```tox black``` warned for ```optuna/integration/xgboost.py``` in my local but CI does not allow the change. So I rolled back it.
```
    black run-test: commands[0] | black . --check --diff
    --- optuna/integration/xgboost.py       2021-12-13 23:34:01.614238 +0000
    +++ optuna/integration/xgboost.py       2021-12-14 00:17:52.997456 +0000
    @@ -72,10 +72,11 @@
                     message = "Trial was pruned at iteration {}.".format(epoch)
                     raise optuna.TrialPruned(message)
                 # The training should not stop.
                 return False
    
    +
     elif _imports.is_successful():
    
         def _get_callback_context(env: "xgb.core.CallbackEnv") -> str:
             """Return whether the current callback context is cv or train.
    
    @@ -111,10 +112,11 @@
                 self._trial.report(current_score, step=env.iteration)
                 if self._trial.should_prune():
                     message = "Trial was pruned at iteration {}.".format(env.iteration)
                     raise optuna.TrialPruned(message)
    
    +
     else:
    
         class XGBoostPruningCallback:  # type: ignore
             __doc__ = _doc
    
    would reformat optuna/integration/xgboost.py
    Oh no! 💥 💔 💥
    1 file would be reformatted, 281 files would be left unchanged.
```

nzw0301 commented:
## `black`

`black` in your environment is supposed to be a little bit old.

masap commented:
The ```black``` in my environment is 21.12b0, the same as CI. I will check my environment. Anyway I fixed my PR, could you review this ?

nzw0301 commented:
I'm not familiar with logging of tqdm, which is used for the progress bar, but I can simplify the code as follows. 

```python
def run_optimize_with_progbar(args: Tuple[str, str, int]) -> str:
    study_name = args[0]
    storage_url = args[1]
    n_jobs = args[2]

    # Create a study
    study = optuna.load_study(study_name=study_name, storage=storage_url)

    tmp = tempfile.NamedTemporaryFile()

    with open(tmp.name, "w") as f:
        with contextlib.redirect_stderr(f):
            study.optimize(
                objective,
                n_trials=20,
                n_jobs=n_jobs,
                show_progress_bar=True,
            )
            f.flush()

    with open(tmp.name, "r") as f:
        return f.read()


@pytest.mark.parametrize("n_jobs", [1, 4])
def test_multiprocess_with_progbar(storage_url: str, n_jobs: int) -> None:
    n_workers = 8
    study_name = _STUDY_NAME
    optuna.create_study(storage=storage_url, study_name=study_name)
    with Pool(n_workers) as pool:
        progress_bars_as_str = pool.map(
            run_optimize_with_progbar, [(study_name, storage_url, n_jobs)] * n_workers
        )

        n_completed_bars = 0
        for progress_bar in progress_bars_as_str:
            if "20/20" in progress_bar and "100%" in progress_bar:
                n_completed_bars += 1

        assert n_completed_bars == n_workers

    study = optuna.load_study(study_name=study_name, storage=storage_url)

    trials = study.trials
    assert len(trials) == n_workers * 20

    _check_trials(trials)

```

What do you think?

masap commented:
Thanks, Fixed. I have thought that we need ```capsys.disabled()``` to redirect the ```stderr```.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
+    study_name = _STUDY_NAME
+    optuna.create_study(storage=storage_url, study_name=study_name)
+    with Pool(n_workers) as pool:
+        results = pool.map(
```
`results` is a little bit too general. How about this?
```suggestion
        progress_bars_as_str = pool.map(
```

nzw0301 replied:
Sorry, this variable was unclear (retuned value is not str). I'll revise the comment above.

masap replied:
Ok, fixed.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
+        results = pool.map(
+            run_optimize_with_progbar, [(study_name, storage_url, n_jobs)] * n_workers
+        )
+        comp_cnt = 0
```
This kinda abbreviation is also a little bit unclear for those who read this code later. How about this?
```suggestion
        n_completed_progbars = 0
```

masap replied:
Ok, fixed.

nzw0301 reviewed (commented):
Awesome. I've left minor comments. Could you take a look at these?

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
+            run_optimize_with_progbar, [(study_name, storage_url, n_jobs)] * n_workers
+        )
+        comp_cnt = 0
+        for result in results:
```
Depending on the returned value by `pool.map`, but similarity `result` is not clear. One alternative might be as follows.

```suggestion
        for progress_bar in progress_bars:
```

masap replied:
Thanks, all was fixed.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -35,6 +37,23 @@ def run_optimize(args: Tuple[str, str]) -> None:
+    study_name = args[0]
+    storage_url = args[1]
+    n_jobs = args[2]
+    # Create a study
```
```suggestion

    # Create a study
```

masap replied:
I don't think that blank line is necessary because existing ```run_optimize``` does not have blank line before comment ```# Create a study```.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -35,6 +37,23 @@ def run_optimize(args: Tuple[str, str]) -> None:
     study.optimize(objective, n_trials=20)
 
 
+def run_optimize_with_progbar(args: Tuple[str, str, int]) -> str:
```
`run_optimize_with_progbar` returns progress bar as `str`, so how about this?

```suggestion
def get_completed_progbar(args: Tuple[str, str, int]) -> str:
```

If there exist a better name, please let me know!

masap replied:
The functionality of the ```run_optimize_with_progbar``` function is not only to return the progress bar, but also to run optimize. In fact, not only the progress bar but also the generated trials are checked subsequently. Therefore, the ```run_optimize_with_progbar``` function is the ```run_optimize``` function plus the ability to return the progress bar, and I think the name is appropriate.

nzw0301 replied:
I respectfully disagree with you. This function aims to get a completed progress bar that can be obtained by only using `optimize` method in Optuna. Since similar method `run_optimize` does return `None`, we expect `run_optimize_with_progbar` does also the same thing with progress bar option; we expect this returns `None`. Therefore the current name can mislead. 

cc: @toshihikoyanase

nzw0301 replied:
In terms of the test of optimisation results, I'm not sure this test is really necessary because the same test without a progress bar is done on another test.

masap replied:
Ok, renamed to ```get_completed_progbar``` and removed check of trials.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -35,6 +37,23 @@ def run_optimize(args: Tuple[str, str]) -> None:
+            # Run optimization
+            study.optimize(objective, n_trials=20, n_jobs=n_jobs, show_progress_bar=True)
+            f.flush()
+    with open(tmp.name, "r") as f:
```
```suggestion

    with open(tmp.name, "r") as f:
```

masap replied:
Ok, fixed.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -35,6 +37,23 @@ def run_optimize(args: Tuple[str, str]) -> None:
+    tmp = tempfile.NamedTemporaryFile()
+    with open(tmp.name, "w") as f:
+        with contextlib.redirect_stderr(f):
+            # Run optimization
```
This sounds obvious:
```suggestion
```

masap replied:
I added this comment because existing ```run_optimize``` has same comment before ```study.optimize```.

nzw0301 replied:
The existing code might not be a ground truth; personally, `run_optimize`'s comment looks also unnecessary.

masap replied:
Ok, removed.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
+    n_workers = 8
+    study_name = _STUDY_NAME
+    optuna.create_study(storage=storage_url, study_name=study_name)
+    with Pool(n_workers) as pool:
```
```suggestion

    with Pool(n_workers) as pool:
```

masap replied:
I don't think that blank line is necessary because existing ```test_multiprocess``` does not have blank line before ```with Pool(n_workers) as pool:```.

nzw0301 commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -115,3 +134,26 @@ def test_multiprocess(storage_url: str) -> None:
+        progress_bars = pool.map(
+            run_optimize_with_progbar, [(study_name, storage_url, n_jobs)] * n_workers
+        )
+        n_completed_progbars = 0
```
```suggestion

        n_completed_progbars = 0
```

masap replied:
Ok, fixed.

nzw0301 reviewed (changes requested):
let me leave additional minor comments for readability.

nzw0301 commented:
We can merge this PR with [the squash option](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests), so please reflect your comments without the force option because we cannot see the difference from the previous commit by a new commit.

nzw0301 commented:
Thank you for your quick update and response! If you are not sure whether or not to reflect my comments, you can wait for another reviewer's comments without hesitation!

masap commented:
> We can merge this PR with the squash option, so please reflect your comments without the force option because we cannot see the difference from the previous commit by a new commit.

I have split this fix as a commit named "Additional fix for PR #3138". I needed to push with force option because I rebased my repo with current master.

nzw0301 commented:
I'll review this PR again after https://github.com/optuna/optuna/pull/3115 is merged into the master branch.

nzw0301 commented:
According to https://github.com/optuna/optuna/pull/3115#discussion_r759996119, `n_jobs != 1` with `timeout`-based progress bar does not work as we expected. So I'll suggest disabling the progress bar if progress bar tries to use `timeout` when `n_jobs != 1` in this PR.

masap commented:
> disabling the progress bar if progress bar tries to use timeout when n_jobs != 1 in this PR.

Ok, fixed.

nzw0301 commented on optuna/study/_optimize.py:
```diff
@@ -56,8 +56,11 @@ def _optimize(
 
-    # TODO(crcrpar): Make progress bar work when n_jobs != 1.
-    progress_bar = pbar_module._ProgressBar(show_progress_bar and n_jobs == 1, n_trials, timeout)
+    if show_progress_bar and timeout is not None and n_jobs != 1:
```
```suggestion
    if show_progress_bar and n_trials is None and timeout is not None and n_jobs != 1:
```

nzw0301 commented on optuna/study/_optimize.py:
```diff
@@ -56,8 +56,11 @@ def _optimize(
 
-    # TODO(crcrpar): Make progress bar work when n_jobs != 1.
-    progress_bar = pbar_module._ProgressBar(show_progress_bar and n_jobs == 1, n_trials, timeout)
+    if show_progress_bar and timeout is not None and n_jobs != 1:
```
When `n_trials` is not `None`, progress bar uses `n_trials` instead of `timeout`:
```suggestion
    if show_progress_bar and n_trials is None and timeout is not None and n_jobs != 1:
```

masap replied:
Ok ,fixed.

nzw0301 commented on optuna/study/_optimize.py:
```diff
@@ -56,8 +56,11 @@ def _optimize(
-    # TODO(crcrpar): Make progress bar work when n_jobs != 1.
-    progress_bar = pbar_module._ProgressBar(show_progress_bar and n_jobs == 1, n_trials, timeout)
+    if show_progress_bar and timeout is not None and n_jobs != 1:
+        warnings.warn("Progress bar is disabled because timeout is enabled and n_jobs != 1.")
```
Like this?
```suggestion
        warnings.warn("The timeout-based progress bar is not supported with n_jobs != 1.")
```

masap replied:
Ok ,fixed.

nzw0301 commented on tests/study_tests/test_study.py:
```diff
@@ -799,9 +799,18 @@ def _objective(trial: Trial) -> float:
 def test_optimize_with_progbar_parallel(capsys: _pytest.capture.CaptureFixture) -> None:
 
     study = create_study()
-    with pytest.warns(UserWarning, match="Progress bar only supports serial execution"):
-        study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=-1)
```
Could you add similar test to this part for `timeout`-based progress bar with `n_jobs!=1`?

masap replied:
Ok ,fixed.

nzw0301 reviewed (approved):
LGTM. Thanks!

toshihikoyanase commented on tests/storages_tests/rdb_tests/test_with_server.py:
```diff
@@ -115,3 +134,20 @@ def test_multiprocess(storage_url: str) -> None:
     _check_trials(trials)
+
+
+@pytest.mark.parametrize("n_jobs", [1, 4])
+def test_multiprocess_with_progbar(storage_url: str, n_jobs: int) -> None:
+    n_workers = 8
+    study_name = _STUDY_NAME
+    optuna.create_study(storage=storage_url, study_name=study_name)
+    with Pool(n_workers) as pool:
+        progress_bars = pool.map(
+            get_completed_progbar, [(study_name, storage_url, n_jobs)] * n_workers
+        )
+
+        n_completed_progbars = 0
+        for progress_bar in progress_bars:
+            if "20/20" in progress_bar and "100%" in progress_bar:
+                n_completed_progbars += 1
+        assert n_completed_progbars == n_workers
```
Can we remove this test case?

1. `tests/storages_tests/rdb_tests/test_with_server.py` has the test cases to check the behavior of `RDBStorage` with multiple RDB servers such as MySQL and PostgreSQL. I don't think the progress bar feature depends on them.

2. The suggested test code executes `study.optimize` in parallel processes,  so each progress bar does not interact with others. So, I think we can simplify this test case by just checking the progress bar in a process.

masap replied:
Ok, removed.

toshihikoyanase replied:
Thanks!

toshihikoyanase commented on tests/study_tests/test_study.py:
```diff
@@ -799,9 +799,21 @@ def _objective(trial: Trial) -> float:
     study = create_study()
-    with pytest.warns(UserWarning, match="Progress bar only supports serial execution"):
-        study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=-1)
+    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=-1)
```
This is not actually a change of this PR, but I think `n_jobs=-1` makes the test environmental dependent. How about setting `n_jobs=2` as can be seen in `test_optimize_with_progbar_parallel_timeout`?
```suggestion
    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=2)
```

masap replied:
Ok, fixed.

toshihikoyanase commented on tests/study_tests/test_study.py:
```diff
@@ -799,9 +799,21 @@ def _objective(trial: Trial) -> float:
+    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=-1)
+    _, err = capsys.readouterr()
+
+    # search for progress bar elements in stderr
```
A nit.
```suggestion
    # Search for progress bar elements in stderr.
```

masap replied:
Ok, fixed both `test_optimize_with_progbar` and `test_optimize_with_progbar_parallel`.

I rebased and squashed these to one commit.

toshihikoyanase replied:
Thank you!

toshihikoyanase reviewed (changes requested):
I'm sorry for the delayed response. The change in `_optimize.py` looks great to me, but I have some comments on the test code.

toshihikoyanase commented on tests/study_tests/test_study.py:
```diff
@@ -799,9 +799,21 @@ def _objective(trial: Trial) -> float:
+    assert "100%" in err
 
+
+def test_optimize_with_progbar_parallel_timeout(capsys: _pytest.capture.CaptureFixture) -> None:
```
I found that the first assertion (i.e., `assert "10/10" not in err` ) in this function is True even if the value of `n_jobs` is 1. How about replacing the assertions with `assert "|" not in err`? It can be seen in `test_optimize_progbar_no_constraints` and it ensures that any progress-bar-related messages are not shown.

```diff
-    assert "10/10" not in err
-    assert "100%" not in err
+    # Testing for a character that forms progress bar borders.
+    assert "|" not in err
```

masap replied:
Right, fixed.

toshihikoyanase replied:
Thanks!

toshihikoyanase commented on tests/study_tests/test_study.py:
```diff
@@ -799,9 +799,21 @@ def _objective(trial: Trial) -> float:
 def test_optimize_with_progbar_parallel(capsys: _pytest.capture.CaptureFixture) -> None:
 
     study = create_study()
-    with pytest.warns(UserWarning, match="Progress bar only supports serial execution"):
-        study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=-1)
+    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=2)
+    _, err = capsys.readouterr()
+
+    # Search for progress bar elements in stderr.
+    assert "10/10" in err
+    assert "100%" in err
```
I found that most of the lines in this test function are duplicated with `test_optimize_with_progbar`. If we apply `parametrize` to `test_optimize_with_progbar` in terms of `n_jobs`, we can reduce the code and improve the maintainability.

```python
@pytest.mark.parametrize("n_jobs", [1, 2])
def test_optimize_with_progbar(n_jobs: int, capsys: _pytest.capture.CaptureFixture) -> None:

    study = create_study()
    study.optimize(lambda _: 1.0, n_trials=10, show_progress_bar=True, n_jobs=n_jobs)
    _, err = capsys.readouterr()

    # Search for progress bar elements in stderr.
    assert "10/10" in err
    assert "100%" in err
```

In addition, I also found some missing test cases, for example, (`n_jobs=2`, `n_trials=10`, `timeout=2`) and (`n_jobs=2`, `n_jobs=None`, `timeout=None`). So, please apply `parametrize` to the following test functions as well.

- `test_optimize_without_progbar`
- `test_optimize_progbar_n_trials_prioritized`
- `test_optimize_progbar_no_constraints`

masap replied:
> I found that most of the lines in this test function are duplicated with test_optimize_with_progbar. If we apply parametrize to test_optimize_with_progbar in terms of n_jobs, we can reduce the code and improve the maintainability.

Ok, fixed.

> In addition, I also found some missing test cases, for example, (n_jobs=2, n_trials=10, timeout=2) and (n_jobs=2, n_jobs=None, timeout=None). So, please apply parametrize to the following test functions as well.

Ok, fixed.

| show_progress_bar | n_trials | timeout | n_jobs | covered by |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| False | None | None | 1 | test_optimize_without_progbar_no_constraints |
| False | None | None | not 1 | test_optimize_without_progbar_no_constraints |
| False | None | not None | 1 | test_optimize_without_progbar_timeout |
| False | None | not None | not 1 | test_optimize_without_progbar_timeout |
| False | not None | None | 1 | test_optimize_without_progbar |
| False | not None | None | not 1 | test_optimize_without_progbar |
| False | not None | not None | 1 | test_optimize_without_progbar_n_trials_prioritized |
| False | not None | not None | not 1 | test_optimize_without_progbar_n_trials_prioritized |
| True | None | None | 1 | test_optimize_progbar_no_constraints |
| True | None | None | not 1 | test_optimize_progbar_no_constraints |
| True | None | not None | 1 | test_optimize_with_progbar_timeout |
| True | None | not None | not 1 | test_optimize_with_progbar_parallel_timeout |
| True | not None | None | 1 | test_optimize_with_progbar |
| True | not None | None | not 1 | test_optimize_with_progbar |
| True | not None | not None | 1 | test_optimize_progbar_n_trials_prioritized |
| True | not None | not None | not 1 | test_optimize_progbar_n_trials_prioritized |

toshihikoyanase reviewed (changes requested):
Thank you for your update! I found some missing test cases, and I guess we can further improve the test coverage.

toshihikoyanase reviewed (approved):
Thank you for summarizing the test case in https://github.com/optuna/optuna/pull/3138#discussion_r790035789. The change looks great to me!

toshihikoyanase commented:
The table in https://github.com/optuna/optuna/pull/3138#discussion_r790035789 may suggest that we can reduce the number of test functions and simplify the test logic. I'm not sure, but we can work on it after discussing the test policy in https://github.com/optuna/optuna/issues/2974.
